storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -55,6 +113,2298 @@ public class UnifiedLog {
+                Optional<LogSegment> nextSegmentOpt = nextOption(segmentsIterator);
+                boolean isLastSegmentAndEmpty = nextSegmentOpt.isEmpty() && segment.size() == 0;
+                long upperBoundOffset = nextSegmentOpt.map(LogSegment::baseOffset).orElseGet(this::logEndOffset);
+                // We don't delete segments with offsets at or beyond the high watermark to ensure that the log start
+                // offset can never exceed it.
+                boolean predicateResult = highWatermark() >= upperBoundOffset && predicate.execute(segment, nextSegmentOpt);
+
+                // Roll the active segment when it breaches the configured retention policy. The rolled segment will be
+                // eligible for deletion and gets removed in the next iteration.
+                if (predicateResult && remoteLogEnabled() && nextSegmentOpt.isEmpty() && segment.size() > 0) {
+                    shouldRoll = true;
+                }
+                if (predicateResult && !isLastSegmentAndEmpty && isSegmentEligibleForDeletion(nextSegmentOpt, upperBoundOffset)) {
+                    deletable.add(segment);
+                    segmentOpt = nextSegmentOpt;
+                } else {
+                    segmentOpt = Optional.empty();
+                }
+            }
+            if (shouldRoll) {
+                logger.info("Rolling the active segment to make it eligible for deletion");
+                roll();
+            }
+            return deletable;
+        }
+    }
+
+    /**
+     * Wraps the value of iterator.next() in an optional.
+     *
+     * @param iterator the iterator
+     * @return Optional of the next element if it exists otherwise Optional.empty()
+     */
+    private static <T> Optional<T> nextOption(Iterator<T> iterator) {
+        return iterator.hasNext()
+            ? Optional.of(iterator.next())
+            : Optional.empty();
+    }
+
+    private int deleteSegments(List<LogSegment> deletable, SegmentDeletionReason reason) {
+        return maybeHandleIOException("Error while deleting segments for " + topicPartition() + " in dir " + dir().getParent(),
+                () -> {
+                    int numToDelete = deletable.size();
+                    if (numToDelete > 0) {
+                        // we must always have at least one segment, so if we are going to delete all the segments, create a new one first
+                        List<LogSegment> segmentsToDelete = deletable;
+                        if (localLog.segments().numberOfSegments() == numToDelete) {
+                            LogSegment newSegment = roll();
+                            if (deletable.get(deletable.size() - 1).baseOffset() == newSegment.baseOffset()) {
+                                logger.warn("Empty active segment at {} was deleted and recreated due to {}", deletable.get(deletable.size() - 1).baseOffset(), reason);
+                                deletable.remove(deletable.size() - 1);
```

Comment:
we should not remove the last segment from the `deletable`, right?

Reply:
I'm a bit uneasy making this type of changes. The current logic matches what was done in Scala and passes all tests. If we find things we want to change we could do them in follow up PRs.

Reply:
> deletable is utilized to remove snapshot files. but it seems to be unnecessary to delete the snapshot associated with the empty active segment, as there should be no corresponding snapshot file.

Good point. The PR changed the implementation a bit. Before this PR, we remove the snapshot associated with the empty segment. With the PR, we keep it. I think the latter is slightly better. Snapshot is the accumulated result from all previous segments. A snapshot is created on every roll. So even an empty active segment could have a non-empty snapshot. Keeping it with an empty segment avoids losing the sequence number and the txn state from the producer clients.

Reply:
>  I think the latter is slightly better. 

agree!

> I'm a bit uneasy making this type of changes. The current logic matches what was done in Scala and passes all tests. If we find things we want to change we could do them in follow up PRs.

I believe we should retain this change, as it appears to be a bug fix.